pkg/util/kubernetes/apiserver/resourcetypes_test.go

```diff
@@ -0,0 +1,258 @@
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/discovery"
+	fakediscovery "k8s.io/client-go/discovery/fake"
+	fakeclientset "k8s.io/client-go/kubernetes/fake"
+)
+
+func TestInitializeGlobalResourceTypeCache(t *testing.T) {
+	resetCache()
+
+	client := fakeclientset.NewClientset()
+	fakeDiscoveryClient := client.Discovery().(*fakediscovery.FakeDiscovery)
+	fakeDiscoveryClient.Resources = []*v1.APIResourceList{}
+
+	tests := []struct {
+		name      string
+		discovery discovery.DiscoveryInterface
+		wantErr   bool
+	}{
+		{
+			name:      "Initialize cache successfully",
+			discovery: fakeDiscoveryClient,
+			wantErr:   false,
+		},
+		{
+			name:      "Re-initialization does nothing",
+			discovery: fakeDiscoveryClient,
+			wantErr:   false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := InitializeGlobalResourceTypeCache(tt.discovery)
+			if tt.wantErr {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestGetResourceType(t *testing.T) {
+	resetCache()
+
+	client := fakeclientset.NewClientset()
+	fakeDiscoveryClient := client.Discovery().(*fakediscovery.FakeDiscovery)
+	fakeDiscoveryClient.Resources = []*v1.APIResourceList{
+		{
+			GroupVersion: "v1",
+			APIResources: []v1.APIResource{
+				{Kind: "Pod", Name: "pods"},
+			},
+		},
+	}
+
+	err := InitializeGlobalResourceTypeCache(fakeDiscoveryClient)
+	assert.NoError(t, err)
+
+	tests := []struct {
+		name    string
+		kind    string
+		group   string
+		want    string
+		wantErr bool
+	}{
+		{
+			name:    "Cache hit for Pod/v1",
+			kind:    "Pod",
+			group:   "v1",
+			want:    "pods",
+			wantErr: false,
+		},
+		{
+			name:    "Cache miss for unknown kind",
+			kind:    "UnknownKind",
+			group:   "unknownGroup",
+			want:    "",
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := GetResourceType(tt.kind, tt.group)
+			if tt.wantErr {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+				assert.Equal(t, tt.want, got)
+			}
+		})
+	}
+}
+
+func TestDiscoverResourceType(t *testing.T) {
+	resetCache()
+
+	client := fakeclientset.NewClientset()
+	fakeDiscoveryClient := client.Discovery().(*fakediscovery.FakeDiscovery)
+	fakeDiscoveryClient.Resources = []*v1.APIResourceList{
+		{
+			GroupVersion: "apps/v1",
+			APIResources: []v1.APIResource{
+				{Kind: "Deployment", Name: "deployments"},
+			},
+		},
+	}
+
+	err := InitializeGlobalResourceTypeCache(fakeDiscoveryClient)
+	assert.NoError(t, err)
+
+	tests := []struct {
+		name    string
+		kind    string
+		group   string
+		want    string
+		wantErr bool
+	}{
+		{
+			name:    "Find Deployment in apps/v1",
+			kind:    "Deployment",
+			group:   "apps",
+			want:    "deployments",
+			wantErr: false,
+		},
+		{
+			name:    "Resource not found",
+			kind:    "UnknownKind",
+			group:   "unknownGroup",
+			want:    "",
+			wantErr: true,
+		},
+	}
```

Comment:
Might be interesting to add some test cases where you have to trim the subresource/where the subresource is invalid